### PR TITLE
fix #1374 remove jars of yarn in executor's classpath

### LINKED ISSUE #1374: When launching an executor, the class path includes yarn jars

### PR DESCRIPTION
**Pull request by huafengw**

### CODE REVIEW

File: daemon/src/main/scala/io/gearpump/cluster/worker/Worker.scala [hunk 1]
```diff
@@ -441,14 +447,17 @@ private[cluster] object Worker {
...
       //The class path maybe a symbol link
-      val canonicalized = Try(new File(path).getCanonicalPath)
-      if(canonicalized.isSuccess && canonicalized.get.startsWith(daemonDir)) {
-        true
-      } else {
+      val filePath = Try(new File(path).getCanonicalPath)
+      val parentPath = Try(new File(path).getParentFile.getCanonicalPath)
+      if((filePath.isSuccess && whiteList.contains(filePath.get)) ||
+          (parentPath.isSuccess && whiteList.contains(parentPath.get))) {
         false
```
Comment (manuzhang) [inline daemon/src/main/scala/io/gearpump/cluster/worker/Worker.scala]: double negative means positive ?

Comment (manuzhang) [inline daemon/src/main/scala/io/gearpump/cluster/worker/Worker.scala]: double negative means positive ?

File: daemon/src/main/scala/io/gearpump/cluster/worker/Worker.scala [hunk 2]
```diff
@@ -269,6 +269,12 @@ private[cluster] class Worker(masterProxy : ActorRef) extends Actor with TimeOut
 }
 
 private[cluster] object Worker {
+  def getClasspathWhiteList(gearHome: String): Array[String] = {
+    Array(
+      gearHome + "/conf",
```
Comment (manuzhang) [inline daemon/src/main/scala/io/gearpump/cluster/worker/Worker.scala]: does the file separator work for windows ?

Comment (huafengw) [inline daemon/src/main/scala/io/gearpump/cluster/worker/Worker.scala]: thx, I'll fix this

File: daemon/src/main/scala/io/gearpump/cluster/worker/Worker.scala [hunk 3]
```diff
@@ -441,11 +447,14 @@ private[cluster] object Worker {
 
   private def filterDaemonLib(classPath: Array[String]): Array[String] = {
     val jarFile = new File(getClass.getProtectionDomain.getCodeSource.getLocation.toURI.getPath)
-    val daemonDir = jarFile.getParent
-    classPath.filterNot{path =>
+    val gearHome = jarFile.getParentFile.getParentFile.getParent
```
Comment (whjiang) [inline daemon/src/main/scala/io/gearpump/cluster/worker/Worker.scala]: better to move this gearHome path finding code into a standalone function and detail document the rationale and assumed dir structure.

File: daemon/src/main/scala/io/gearpump/cluster/worker/Worker.scala [hunk 4]
```diff
@@ -437,18 +439,19 @@ private[cluster] object Worker {
...
-        true
-      } else {
-        false
+    private def filterDaemonLib(classPath: Array[String]): Array[String] = {
+      val classpathWhiteList = getClasspathWhiteList(config)
+      classPath.filter{ path =>
+        //The class path maybe a symbol link
+        val filePath = Try(new File(path).getCanonicalPath)
+        val parentPath = Try(new File(path).getParentFile.getCanonicalPath)
+        if((filePath.isSuccess && classpathWhiteList.contains(filePath.get)) ||
```
Comment (clockfly) [inline daemon/src/main/scala/io/gearpump/cluster/worker/Worker.scala]: What filter here mean?
What filePath.isSuccess mean?
This function is not clear about what it try to do.

Comment (huafengw) [inline daemon/src/main/scala/io/gearpump/cluster/worker/Worker.scala]: To resolve the symbol link, function ```getCanoncialPath``` may throw exception.

Comment (clockfly) [inline daemon/src/main/scala/io/gearpump/cluster/worker/Worker.scala]: And what that means?

Comment (huafengw) [inline daemon/src/main/scala/io/gearpump/cluster/worker/Worker.scala]: The class path may be a symbol link, resolve it and check whether it's in the white list. The class path also can be a folder or file, so need to check itself and its' parent

Comment (clockfly) [inline daemon/src/main/scala/io/gearpump/cluster/worker/Worker.scala]: What if it is a symbol link? getCanoncialPath will throw?

Please refactor this to make it more clear.

**Comment by clockfly** [comment]
Why you need to care whether it is a symbolic link or not?

File: daemon/src/main/scala/io/gearpump/cluster/worker/Worker.scala [hunk 5]
```diff
@@ -437,18 +439,23 @@ private[cluster] object Worker {
...
+
+      def inWhiteList(file: File): Boolean = {
+        //Resolve the file path's symbol link, the function may throw exception when resolving failed
+        val canonicalPath = Try(file.getCanonicalPath)
+        canonicalPath.isSuccess && classpathWhiteList.contains(canonicalPath.get)
+      }
+
+      classPath.filter{ path =>
+        //The path may represent a file or folder
+        if(inWhiteList(new File(path)) || inWhiteList(new File(path).getParentFile)) {
```
Comment (stanleyxu2005) [inline daemon/src/main/scala/io/gearpump/cluster/worker/Worker.scala]: The expression itself indicates a boolean value, no need to return true/false explicitly.

**Comment by huafengw** [comment]
Fixed

File: daemon/src/main/scala/io/gearpump/cluster/worker/Worker.scala [hunk 6]
```diff
@@ -437,18 +438,13 @@ private[cluster] object Worker {
...
-    val daemonDir = jarFile.getParent
-    classPath.filterNot{path =>
-      //The class path maybe a symbol link
-      val canonicalized = Try(new File(path).getCanonicalPath)
-      if(canonicalized.isSuccess && canonicalized.get.startsWith(daemonDir)) {
-        true
-      } else {
-        false
+    private def filterDaemonLib(classPath: Array[String]): Array[String] = {
+      val classpathWhiteList = getClasspathWhiteList(config)
```
Comment (stanleyxu2005) [inline daemon/src/main/scala/io/gearpump/cluster/worker/Worker.scala]: Change all `classpath` to `classPath`; all `Classpath` to `ClassPath`.

File: daemon/src/main/scala/io/gearpump/cluster/worker/Worker.scala [hunk 7]
```diff
@@ -437,18 +438,13 @@ private[cluster] object Worker {
...
+      classPath.filter{ path =>
+        //The path may represent a file or folder
+        classpathWhiteList.contains(new File(path).getAbsolutePath) ||
```
Comment (stanleyxu2005) [inline daemon/src/main/scala/io/gearpump/cluster/worker/Worker.scala]: Here might call `new File(path)` twice. Create an object first.

**Comment by clockfly** [comment]
Have you tested on different OS?

Sometimes classpath has path prefix like file:///

**Comment by huafengw** [comment]
Works fine on Windows

**Comment by clockfly** [comment]
+1

**Pull request merged**